### PR TITLE
kv: introduce a stopgap for lack of ReplicaState synchronization

### PR DESCRIPTION
**Pull request by irfansharif**
There's a scary lack of synchronization around how we set the
ReplicaState for a given replica, and how we mark a replica as
"initialized". What this means is that very temporarily, it's possible
for the entry in Store.mu.replicas to be both "initialized" and have an
empty ReplicaState. This was an existing problem, but is now more likely
to bite us given the migrations infrastructure attempts to purge
outdated replicas at start up time (when replicas are being initialized,
and we're iterating through extan replicas in the Store.mu.replicas
map).

This issue has caused a bit of recent instability: #59180, #58489,
\#58523, and #58378. While we work on a more considered fix to the
problem (tracked in #58489), we can introduce stop the bleeding in the
interim (and unskip some tests).

Release note: None

### CODE REVIEW

**Review by tbg — APPROVED** [review]
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59194)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

**Comment by irfansharif** [comment]
> curl: (22) The requested URL returned error: 503 Service Unavailable: Back-end server is at capacity

Blocked on CI like everyone else. TFTR! Will merge after things settle down.

**Pull request merged**
